### PR TITLE
Rework the API expressed by DgsExecutionResult

### PR DESCRIPTION
**Pull request by berngp**
Pull Request type
----

- [ ] Bugfix
- [ ] Feature
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe):

Changes in this PR
----

* We are adopting a Builder pattern instead of depending on Kotlin delegation for the ExecutionResult. This was done to expose a cleaner API for Java Consumers.
* We are also moving the `DgsExecutionResult` to not be _internal_, since at is clear from the `MyInstrumentation` example developers will be using it explicitly.

### CODE REVIEW

File: graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/DgsExecutionResult.kt [hunk 1]
```diff
@@ -133,4 +86,56 @@ class DgsExecutionResult @JvmOverloads constructor(
...
+     * Facilitate the construction of a [DgsExecutionResult] instance.
+     */
+    class Builder {
+        var executionResult: ExecutionResult = DEFAULT_EXECUTION_RESULT
+            private set
+        fun executionResult(executionResult: ExecutionResult) = apply { this.executionResult = executionResult }
+
+        fun executionResult(executionResultBuilder: ExecutionResultImpl.Builder) = apply { this.executionResult = executionResultBuilder.build() }
+
+        var headers: HttpHeaders = HttpHeaders()
```
Comment (kilink) [inline graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/DgsExecutionResult.kt:121]: Can this default to `HttpHeaders.EMPTY`, so it doesn't need to instantiate anything if the builder method isn't called?

Comment (berngp) [reply graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/DgsExecutionResult.kt:121]: It should, thanks @kilink !

File: graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/DgsExecutionResult.kt [hunk 2]
```diff
@@ -14,86 +14,39 @@
...
 import org.springframework.http.HttpHeaders
 import org.springframework.http.HttpStatus
 import org.springframework.http.ResponseEntity
 
-class DgsExecutionResult @JvmOverloads constructor(
+class DgsExecutionResult(
     val executionResult: ExecutionResult,
-    val headers: HttpHeaders = HttpHeaders(),
-    val status: HttpStatus = HttpStatus.OK
+    val headers: HttpHeaders,
```
Comment (kilink) [inline graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/DgsExecutionResult.kt:34]: `HttpHeaders` is exposed via a public getter, should we wrap it with `HttpHeaders.readOnlyHttpHeaders` so we're not exposing mutable state?

File: graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/DgsExecutionResult.kt [hunk 3]
```diff
@@ -14,86 +14,39 @@
...
 import graphql.GraphqlErrorBuilder
 import org.slf4j.Logger
 import org.slf4j.LoggerFactory
 import org.springframework.http.HttpHeaders
 import org.springframework.http.HttpStatus
 import org.springframework.http.ResponseEntity
 
-class DgsExecutionResult @JvmOverloads constructor(
+class DgsExecutionResult(
     val executionResult: ExecutionResult,
```
Comment (kilink) [inline graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/DgsExecutionResult.kt:33]: I wonder if this property needs to be public, we are already delegating all methods to it

**Review by kilink — APPROVED** [review]

**Review by srinivasankavitha — APPROVED** [review]

**Comment by antholeole** [comment]
Ah, I didn't consider these things, sorry! These are good changes, thanks for the refactor.

**Pull request merged**